From cd870d90fd3dd23796464251323ffbbaa4d06669 Mon Sep 17 00:00:00 2001
From: Jan Nieuwenhuizen <janneke@gnu.org>
Date: Tue, 31 Mar 2009 15:35:26 +0200
Subject: [PATCH] OOXML: parse <a:hlinkClick> in <xdr:sp> (shapecontext).  Fixes n#480854.

Hyperlinks can apparently be set in ooxml in several ways.  If
an hyperlink is specified as an external relation with a textbox
shape inside a drawing, this link should be set on the textbox.

   * Modified     oox/inc/oox/drawingml/shapecontext.hxx
   * Modified     oox/source/drawingml/shapecontext.cxx
---
 oox/inc/oox/drawingml/shapecontext.hxx |    1 +
 oox/source/drawingml/shapecontext.cxx  |   10 ++++++++++
 2 files changed, 11 insertions(+), 0 deletions(-)

diff --git oox/inc/oox/drawingml/shapecontext.hxx oox/inc/oox/drawingml/shapecontext.hxx
index 815f808..3236876 100644
--- oox/inc/oox/drawingml/shapecontext.hxx
+++ oox/inc/oox/drawingml/shapecontext.hxx
@@ -54,6 +54,7 @@ protected:
 
 	ShapePtr mpMasterShapePtr;
 	ShapePtr mpShapePtr;
+    rtl::OUString maHyperLink;
 };
 
 } }
diff --git oox/source/drawingml/shapecontext.cxx oox/source/drawingml/shapecontext.cxx
index 6b94270..a678a39 100644
--- oox/source/drawingml/shapecontext.cxx
+++ oox/source/drawingml/shapecontext.cxx
@@ -42,6 +42,7 @@
 #include "oox/drawingml/drawingmltypes.hxx"
 #include "oox/drawingml/customshapegeometry.hxx"
 #include "oox/drawingml/textbodycontext.hxx"
+#include "oox/core/xmlfilterbase.hxx"
 #include "tokens.hxx"
 
 using rtl::OUString;
@@ -91,6 +92,13 @@ Reference< XFastContextHandler > ShapeContext::createFastChildContext( sal_Int32
 		mpShapePtr->setId( xAttribs->getOptionalValue( XML_id ) );
 		mpShapePtr->setName( xAttribs->getOptionalValue( XML_name ) );
 		break;
+        // let upstream do C-x h M-x indent-region
+        case XML_hlinkClick:
+        {
+            OUString aRelId = xAttribs->getOptionalValue( NMSP_RELATIONSHIPS|XML_id );
+            maHyperLink = getRelations().getTargetFromRelId( aRelId );
+            break;
+        }
 	case XML_ph:
 		mpShapePtr->setSubType( xAttribs->getOptionalValueToken( XML_type, XML_obj ) );
 		mpShapePtr->setIndex( xAttribs->getOptionalValue( XML_idx ).toInt32() );
@@ -108,6 +116,8 @@ Reference< XFastContextHandler > ShapeContext::createFastChildContext( sal_Int32
 	case XML_txBody:
 	{
         TextBodyPtr xTextBody( new TextBody );
+        if (maHyperLink.getLength ())
+            xTextBody->getTextProperties ().maPropertyMap[CREATE_OUSTRING ("Hyperlink")] <<= maHyperLink;
         mpShapePtr->setTextBody( xTextBody );
         xRet = new TextBodyContext( *this, *xTextBody );
 		break;
-- 
1.6.0.rc1.49.g98a8

